### PR TITLE
[DO_NOT_MERGE] Use FreeBSD compatibility definition for ElfW() macro on FreeBSD.

### PR DESCRIPTION
**Pull request by yesudeep**
Addresses: https://github.com/abseil/abseil-cpp/issues/1024

The `ElfW()` macro was added to `elf_generic.h` recently and as such may be absent most current releases of FreeBSD.

See: https://github.com/freebsd/freebsd-src/blob/34e7e4b6a059eee5e4e3e34de5b9d5f0d6e589f9/sys/sys/elf_generic.h#L60

There are other build issues on FreeBSD. I'll address them soonish in other PRs.

### CODE REVIEW

**Comment by yesudeep** [comment]
@derekmauro I've closed my previous PR (https://github.com/abseil/abseil-cpp/pull/1025) as this supercedes it. Could you please take a look? Thank you!

File: absl/debugging/internal/link.h [hunk 1]
```diff
@@ -0,0 +1,24 @@
...
+// limitations under the License.
+
+#ifndef ABSL_DEBUGGING_INTERNAL_LINK_H_
+#define ABSL_DEBUGGING_INTERNAL_LINK_H_
+
+#include <link.h>  // For ElfW() macro.
+
+#if defined(__FreeBSD__) && !defined(ElfW)
+#define ElfW(x) __ElfN(x)
+#endif
```
Comment (derekmauro) [inline absl/debugging/internal/link.h]: I don't think we need a new header just for this. It is fine to just define it in the few places where it is needed.

**Review by derekmauro — CHANGES_REQUESTED** [review]
I'd also like @EmployedRussian to review this.

File: absl/debugging/internal/link.h (hunk 1 above)
Comment (EmployedRussian) [reply absl/debugging/internal/link.h]: I agree with @derekmauro  -- this patch seems like an overkill.
1. It's a temporary fix until FreeBSD commit 34e7e4b6a059e makes it out.
2. There is little danger of polluting the namespace with `ElfW` and having unexpected side effects -- Linux versions already have that macro defined, and FreeBSD ones will have soon.
3. It's unlikely that additional macros will need to be added to `absl/debugging/internal/link.h`.

Comment (yesudeep) [reply absl/debugging/internal/link.h]: Sure done.

File: absl/debugging/internal/elf_mem_image.h [hunk 2]
```diff
@@ -38,7 +38,11 @@
 
 #ifdef ABSL_HAVE_ELF_MEM_IMAGE
 
-#include <link.h>  // for ElfW
+#include <link.h>  // for ElfW() macro.
```
Comment (EmployedRussian) [inline absl/debugging/internal/elf_mem_image.h]: I suggest avoiding superficial changes (like this one).

File: absl/debugging/internal/symbolize.h [hunk 3]
```diff
@@ -28,15 +28,20 @@
 
 #ifdef ABSL_INTERNAL_HAVE_ELF_SYMBOLIZE
 #error ABSL_INTERNAL_HAVE_ELF_SYMBOLIZE cannot be directly set
-#elif defined(__ELF__) && defined(__GLIBC__) && !defined(__native_client__) \
-      && !defined(__asmjs__) && !defined(__wasm__)
+#elif defined(__ELF__) && defined(__GLIBC__) && !defined(__native_client__) && \
+    !defined(__asmjs__) && !defined(__wasm__)
```
Comment (EmployedRussian) [inline absl/debugging/internal/symbolize.h:32]: Likewise.

File: absl/debugging/internal/symbolize.h [hunk 4]
```diff
@@ -28,15 +28,20 @@
...
 #define ABSL_INTERNAL_HAVE_ELF_SYMBOLIZE 1
 
 #include <elf.h>
-#include <link.h>  // For ElfW() macro.
+#include <link.h>  // for ElfW() macro
```
Comment (EmployedRussian) [inline absl/debugging/internal/symbolize.h]: Likewise.

File: absl/debugging/internal/vdso_support.cc [hunk 5]
```diff
@@ -18,7 +18,7 @@
 
 #include "absl/debugging/internal/vdso_support.h"
 
-#ifdef ABSL_HAVE_VDSO_SUPPORT     // defined in vdso_support.h
+#ifdef ABSL_HAVE_VDSO_SUPPORT  // defined in vdso_support.h
```
Comment (EmployedRussian) [inline absl/debugging/internal/vdso_support.cc:22]: Likewise.

File: absl/debugging/internal/vdso_support.cc [hunk 6]
```diff
@@ -50,6 +50,18 @@
 #define AT_SYSINFO_EHDR 33  // for crosstoolv10
 #endif
 
+#if defined(__FreeBSD__)
+
+#if !defined(Elf64_aux_t)
+using Elf64_aux_t = Elf64_Auxinfo;
```
Comment (EmployedRussian) [inline absl/debugging/internal/vdso_support.cc]: Did you mean `auxv_t` ?

This isn't going to work -- it's not `#define`d. On a Linux system we have:
```
# 1135 "/usr/include/elf.h" 3 4
typedef struct
{
  uint32_t a_type;
  union
    {
      uint32_t a_val;
    } a_un;
} Elf32_auxv_t;

typedef struct
{
  uint64_t a_type;
  union
    {
      uint64_t a_val;
    } a_un;
} Elf64_auxv_t;
```
So I think you need:
```
#if defined(__FreeBSD__)
using Elf32_auxv_t = Elf32_Auxinfo;
using Elf64_auxv_t = Elf64_Auxinfo;
#endif 
```

File: absl/debugging/internal/vdso_support.cc [hunk 7]
```diff
@@ -135,9 +147,7 @@ const void *VDSOSupport::SetBase(const void *base) {
   return old_base;
 }
 
-bool VDSOSupport::LookupSymbol(const char *name,
-                               const char *version,
-                               int type,
+bool VDSOSupport::LookupSymbol(const char *name, const char *version, int type,
                                SymbolInfo *info) const {
   return image_.LookupSymbol(name, version, type, info);
```
Comment (EmployedRussian) [inline absl/debugging/internal/vdso_support.cc]: Avoid superficial formatting changes.

File: absl/debugging/symbolize_elf.inc [hunk 8]
```diff
@@ -50,8 +50,8 @@
 
 #include <dlfcn.h>
 #include <elf.h>
-#include <fcntl.h>
 #include <link.h>  // For ElfW() macro.
+#include <fcntl.h>
```
Comment (EmployedRussian) [inline absl/debugging/symbolize_elf.inc:54]: Likewise.

File: absl/debugging/internal/vdso_support.cc (hunk 5 above)
Comment (yesudeep) [reply absl/debugging/internal/vdso_support.cc:22]: I think clang-format is reading the .clang-format file in the repository and auto-formatting. It's a pain to undo all of the autoformatting... may be I'll send a cleaner pr with clang-format disabled.

**Pull request closed without merging**